Feat sync settings, Update latest release to master

Zainrax opened the pull request.

mjs commented on public/javascripts/includes/tags.js:
```diff
@@ -192,6 +206,21 @@ tags.quickNew = function(animal) {
+    let tag = {};
+    tag.event = "false positive";
+    tag.confidence = 0.6;
+    console.log(tags.parseSelect('tagEventInput'));
```
This can probably go

arthurmcgregor replied:
We need to do something about false positives, otherwise it submits an animal with name "-"

Happy to hide the confirm button for false positives if it is something you will never use.

mjs replied:
I just meant that the `console.log` call was probably not required.

Confirming a false positive is still something that's useful.

mjs commented on public/javascripts/includes/tags.js:
```diff
@@ -192,6 +206,21 @@ tags.quickNew = function(animal) {
+    // false positive
+    let tag = {};
+    tag.event = "false positive";
+    tag.confidence = 0.6;
```
Where did 0.6 come from?

arthurmcgregor replied:
The 0.6 is the confidence set with the quickNew function (see line 205) so I used that value again. The quickNew function is called every time you push one of the button categories which show just below the video player, as opposed to pushing the big + to get the full add observation screen.

mjs replied:
That's probably fine then. Alternatively, could you use the same confidence as the machine learning generated tag?

arthurmcgregor replied:
I'll leave it as 0.6 because all the other animal confirmations will have that setting (and I don't want to re-write quickNew()

mjs commented:
This looks good to go but needs to be updated wrt the master branch.

hardiesoft committed:
- Updated bootstrap-vue-next lib
- Updated Vue3
- Reworking mask regions component to disallow self-intersecting regions, and allow an apron so regions can be created slightly outside the camera frame.
- Creating wizard steps for manual uploads
- Moving device status out of modal

hardiesoft committed:
- Fix mask regions tests

hardiesoft committed:
- Fix mask regions tests (2, linting)

hardiesoft committed:
- Fix sunset/sunrise in visits breakdown
- Add more info to mobile device cards in DevicesView

hardiesoft committed:
- Fix reference photo component, make it work on mobile.
- More responsive layouts for device setup.

hardiesoft committed:
- Reference photos are properly scaled when overlaying on CptvPlayer on mobile.
- Overflowing tab list component to make longer tabbed nav lists work better on mobile.

hardiesoft committed:
- Trail cameras use an appropriate daytime image as an automatic reference image.
- Masked regions are now used when adding tracks via the API to ignore tracks that are exclusively inside the masked region.
- Use common track rect drawing code for showing tracks for trailcam images.

hardiesoft committed:
- Added a dummy track to trailcam recordings where there is no track – the actual track is created the first time a user classification is added.  This could be extended to empty CPTV files.

hardiesoft committed:
- Styling dummy track

hardiesoft committed:
- Fix tests

hardiesoft committed:
- Added ability to add a dummy track to a CPTV file that has no tracks.
- Automatically add "missed track" tag to CPTV recordings that get manual IDs when they had no tracks.
- Poll every 30 seconds when an open recording is still processing, and update UI when processing is complete.

hardiesoft committed:
- Improved behaviour when mashing on next/prev buttons in CPTV player very quickly :)

hardiesoft committed:
- Fix null settings read in mask-regions API

hardiesoft committed:
- Remove stations wording on Dashboard

hardiesoft committed:
- API Documentation fix
- Add "dummy-track" to tracker version to make it available to processing.

hardiesoft committed:
- Don't send algorithm with tracks, use default

hardiesoft committed:
- Fix mask-regions api read

hardiesoft committed:
- Clone rather than always update in place DeviceHistory entries for adding reference images and mask regions in the same location.
- Fix route for Visits modal from activity view to include selected track.

hardiesoft committed:
- Clone rather than always update in place DeviceHistory entries for adding reference images and mask regions in the same location.
- Fix route for Visits modal from activity view to include selected track. (2)

hardiesoft committed:
- Improved responsive layout for activity search
- Make search url query driven (WIP)

hardiesoft committed:
- Improved responsive layout for activity search
- Make search url query driven (WIP)
- Fix mask-regions api

hardiesoft committed:
- Improved responsive layout for activity search
- Make search url query driven
- Hack in support for "Animal" class from Megadetector
- Show processing progress for visits in API
- Reflect processing progress in the UI.

hardiesoft committed:
- Add UI to visits when recordings are still processing
- Handle scaling DPI when drawing track boxes on trailcam images

hardiesoft committed:
- Make sure player inherits query params from parent component
- Make recording/visits exports work.

hardiesoft committed:
- Don't delete previous reference images

hardiesoft committed:
Scale UP:
- New bulk querying/reporting API which is *many*, *many* times faster for tag filtering queries over longer periods, esp. when the number of recordings in a project grows.
- Thumbnail images on the front-end only get queried when visible.

hardiesoft committed:
Scale UP:
- CI Tests for new bulk tag mode.

hardiesoft committed:
Merge branch 'main' of https://github.com/TheCacophonyProject/cacophony-web into browse-next-upgrades

# Conflicts:
#	api/api/V1/Device.ts
#	api/api/V1/Recording.ts
#	api/api/V1/recordingUtil.ts
#	api/models/index.ts
#	api/package-lock.json
#	browse-next/src/api/Recording.ts
#	browse-next/src/components/ActivitySearchDescription.vue
#	browse-next/src/components/ImageLoader.vue
#	browse-next/src/components/RecordingsList.vue
#	browse-next/src/components/VisitsDailyBreakdown.vue
#	browse-next/src/views/ActivitySearchView.vue

hardiesoft committed:
Merge branch 'main' of https://github.com/TheCacophonyProject/cacophony-web into browse-next-upgrades

hardiesoft committed:
Merge main

hardiesoft committed:
Merge branch 'main' of https://github.com/TheCacophonyProject/cacophony-web into browse-next-upgrades

hardiesoft committed:
Merge main, add all tagModes to improved bulk query.

hardiesoft committed:
Address review feedback.

hardiesoft committed:
Small fixes

hardiesoft committed:
Small fixes (2)

hardiesoft committed:
Small fixes (3)

hardiesoft committed:
Small fixes (3)

hardiesoft committed:
Small fixes (4)

hardiesoft committed:
Get rid of legacy bulk recordings query in browse-next

hardiesoft committed:
Rework bulk-query, fixing edge-cases and adding more comprehensive tests.
Improve UI, lazy loading of activity page.

hardiesoft committed:
Merge branch 'main' of https://github.com/TheCacophonyProject/cacophony-web into browse-next-upgrades

# Conflicts:
#	api/api/V1/Recording.ts
#	api/classifications/classification.json
#	browse-next/src/views/ActivitySearchView.vue
#	integration-tests/cypress/commands/api/recording.ts
#	integration-tests/cypress/e2e/api/recordings/recording_query_v2.ts

hardiesoft committed:
Fix tests

hardiesoft committed:
Fix tests (2)

hardiesoft committed:
Make sure we return recordings in the correct order.

hardiesoft committed:
Auth.

hardiesoft committed:
Make sure we send recording type when querying!

hardiesoft committed:
Auth again

hardiesoft committed:
Fix recordings export report

hardiesoft committed:
Export larger pages to increase export speed, increase
limit on server-side

hardiesoft committed:
Fix issue where we try to request things from inactive devices and get 403 rejected

hardiesoft committed:
Make sure deleted recordings are not included by default in new bulk queries, and add a test for that.
Display "leporadae" as "rabbit or hare"

hardiesoft committed:
Make sure deleted recordings are not included by default in new bulk queries, and add a test for that.
Display "leporadae" as "rabbit or hare"

Commit:
add recording window and low power mode settings

Commit:
fix lint

Commit:
fix for device extract, move settings up

Commit:
readd location check

hardiesoft committed:
Misc fixes and improvements, starting on non-paginated visits API.

hardiesoft committed:
Linting

hardiesoft committed:
Fix tests

hardiesoft committed:
Fix tests (2)

hardiesoft committed:
Fix error event logging

hardiesoft committed:
Fix cron script permissions for platform usage report

hardiesoft committed:
Update docker-compose.yml

hardiesoft committed:
Add alert relationship for project-wide alerts

hardiesoft committed:
Link to device recordings search from device page

hardiesoft committed:
Link to device recordings search from device page
- Fix earliest time detection for location recording results

hardiesoft committed:
Add ability for the user to manage their own email alerts via browse-next UI.

hardiesoft committed:
Linting

hardiesoft committed:
Fix tests

hardiesoft committed:
Improve device -> recordings button appearance on mobile.

hardiesoft committed:
Improve service report email, improve alert email

hardiesoft committed:
Service email should only be for the last 24 hours

hardiesoft committed:
Improve formatting of service errors.
Make audio recordings export include cacophony index column.

hardiesoft committed:
Make stopped alerts work for thermal devices we didn't hear from in over 24 hours.
Improve formatting of service error emails (stop gmail stripping things?)
Fix email link for new-style alert emails.

hardiesoft committed:
Fix latest recording times update for groups.

hardiesoft committed:
Fix latest recording times update for groups. (2)

hardiesoft committed:
Update stopped device email text

hardiesoft committed:
Update stopped device email text (2)

hardiesoft committed:
Revamp stopped devices notification email.

hardiesoft committed:
Skip heartbeat tests (no longer rely on heartbeats for stops)

hardiesoft committed:
Make alert animal image link to recording.

hardiesoft committed:
Tweak stopped devices logic

hardiesoft committed:
Tweak stopped devices logic again

hardiesoft committed:
Fleshing out daily/weekly digest emails + user preferences.

Zainrax committed:
Merge tag 'v1.30.9' into feat-sync-settings

Zainrax committed:
add fox tag

hardiesoft committed:
Daily+weekly digest emails.
Store user timezone info

hardiesoft committed:
Add missing tag icons.

hardiesoft committed:
Add missing tag icons. (2)

hardiesoft committed:
Add fox tag

hardiesoft committed:
Add ability to hide false-trigger tracks in Recording View.

hardiesoft committed:
White outlines around email tag classification icons so that they show up in dark-mode contexts.

hardiesoft committed:
- Improve thumbnail selection in recordings view.
- Fix stopped devices query
- Add aussie tags to classifications.

hardiesoft committed:
- Diagnostics

Zainrax committed:
fix device settings

Zainrax committed:
Merge tag 'v1.30.14' into feat-sync-settings

Zainrax committed:
fix time format in settings

hardiesoft committed:
- Adjust stopped device query

Zainrax committed:
Merge remote-tracking branch 'origin/visits-api-upgrade' into feat-sync-settings

hardiesoft committed:
- Add ability for super-users to view any group they're not already part of (as an admin user)

hardiesoft committed:
- Implement deletion of recordings in browse-next and tweak deletion API to update book-keeping for group and stations lastRecordingTimes.

hardiesoft committed:
- Modify locations query to be able to return only locations that don't have zero recordings associated with them, so that deleting the last recording in a location allows that location to be hidden.

hardiesoft committed:
- Do latest time fixups for bulkDelete API too.

hardiesoft committed:
- Fix tooltips for activity search

hardiesoft committed:
- Fix local saving of view false-triggers setting when viewing as a super-user

hardiesoft committed:
- Fix hiding of AI false-trigger tags for tracks where there is a user non-false-trigger tag.

hardiesoft committed:
- Fix requests for tracks that are user-created and don't have thumbnails.

hardiesoft committed:
- Fix user settings persistence

hardiesoft committed:
- Fix issue with removing track tags twice.

hardiesoft committed:
- Cosmetic tweak – images that fail to load.

hardiesoft committed:
- Fix thumbnail selection code.

hardiesoft committed:
- Fix display of sunset/sunrise times in visits breakdown for 24/7 cameras.

hardiesoft committed:
- Fix date offsets for digests

hardiesoft committed:
- Fix date offsets for digests

hardiesoft committed:
- Fix date offsets for digests

hardiesoft committed:
- Fix date offsets for digests (2)

hardiesoft committed:
- Remove beta version wording, add additional filter events

hardiesoft committed:
- Fixes for digest email timings and thumbnail selection.

hardiesoft committed:
- Allow viewing live thumbnails and cptv recordings in local dev mode.
- Add track replay button to tagging view.
- Hide audio mode stuff for now.

hardiesoft committed:
- Fix unitVersion grouping in daily service error emails.

hardiesoft committed:
- Linting

hardiesoft committed:
- Additional options for finding a project to view as a super-user

Zainrax committed:
fix to make update device settings invariant

Zainrax committed:
Merge tag 'v1.31.2' into feat-sync-settings

Zainrax committed:
only update if changed

hardiesoft committed:
- Complete reset password flow for browse-next.

hardiesoft committed:
- Complete reset password flow for browse-next.
- Add cypress ui tests for reset password flow.

hardiesoft committed:
- Make project switcher work better
- Properly reset activity search when projects are switched (WIP)

hardiesoft committed:
- Fix linting errors

hardiesoft committed:
- Fix broken activity lists
- Fix weirdness with project switching (when on activity page)

hardiesoft committed:
- Fix CPTV player resizing when browser window is very short vertically, but has a desktop width breakpoint.
- Fix date field not being populated when on activity page and page refreshed.

hardiesoft committed:
- Fix type import

hardiesoft commented:
@Zainrax Is this still a draft?

hardiesoft committed:
- Updated device view, device "dashboard" WIP

hardiesoft committed:
- Updated device view, device "dashboard" WIP (2)

hardiesoft committed:
- Fix visitContext filtering on dashboard.

hardiesoft committed:
- Fix recording window length calulation for absolute recordindg windows.
- Fix display of visits activity search description if a device was previously filtered on.

hardiesoft committed:
- In daily/weekly digest emails, try to fall back to another parent class icon if species icon for tag is not found.

hardiesoft committed:
- Improve thumbnail selection for recordings list.

hardiesoft committed:
- Improve pathological cases of recording search where there are many many recordings in a project over a long time span, and the user wants to search over "all time".  To handle this case, start searching over a max time window of 3 months on the API side, and progressively widen that search until there are no results left or the limit has been reached.

hardiesoft committed:
- Handle recordings export where there are (legacy) recordings without a location.

hardiesoft committed:
- Abort in-flight requests if multiple searches are done in quick succession to keep state correct.
- Script to fixup latest recording times
- Fix locations table view
- Fix incorrect thumbnail selection in recordings activity search
- Don't sync activity search context when we enter a recording.
- Don't try to load tag-icons in track tagger view unless the tagging interface is actually expanded.

hardiesoft committed:
- Fix station-related tests to not use status-recording length test recordings
- Add missing recording labels.

hardiesoft committed:
- When we delete the last recording that isn't a status recording, properly fixup last recording times.

hardiesoft committed:
- Handle aborted activity search queries more gracefully in the UI.

hardiesoft committed:
- Handle aborted activity search queries more gracefully in the UI. (2)

hardiesoft committed:
- Fix recording export length truncation when using non-advanced export

hardiesoft committed:
- Fix recording export length truncation when using non-advanced export (2)

Zainrax committed:
add device sync-settings to browse-next

hardiesoft committed:
- Fixes to dashboard showing non-predator visits and locations
- Fix for unselect locations in dashboard
- Remove unused dashboard view options.
- Add a 'time-sensitive' param to the recordings query, to return from slower queries after a time-out, as long as there are some results gathered.  This will make the UI appear more responsive for sparse queries over longer time periods.
- Add ability to go to activity for locations from locations page for both visits and recordings scoped to the given location.

hardiesoft committed:
- Make max-zoom on maps be not quite as zoomed in (15), since sometimes tiles aren't available for the max zoom level of 16.

hardiesoft committed:
- Tidy locations page some more

Zainrax committed:
merge and add location check on settings endpoint

Zainrax committed:
fix custom recording windows

Zainrax committed:
Merge commit '1e31f23' into feat-sync-settings

hardiesoft committed:
- Change page size calculation for recordings search.

hardiesoft committed:
Merge branch 'visits-api-upgrade' of https://github.com/TheCacophonyProject/cacophony-web into feat-sync-settings

hardiesoft commented on api/models/DeviceHistory.ts:
```diff
@@ -113,5 +121,93 @@ export default function (
+
+  let changed = false;
+  for (const key in incomingSettings) {
+    if (incomingSettings.hasOwnProperty(key)) {
```
Small nit: when is this ever not true, given that we're iterating keys in `incomingSettings`.  In the case where you're worried about inherited keys from up the prototype chain you can use `Object.keys` which doesn't iterate inherited keys, only ones that are `hasOwnProperty`.

hardiesoft commented on types/api/device.d.ts:
```diff
@@ -51,10 +51,25 @@ export interface ApiDeviceLocationFixup {
+export type WindowsSettings = {
+  startRecording: string;
+  stopRecording: string;
+  powerOn: string;
```
`powerOn` and `powerOff` fields are ignored in DOC AI cameras, so not sure if we should include these at all.  I don't know that we'll ever go back to using them.

hardiesoft commented on types/api/device.d.ts:
```diff
@@ -51,10 +51,25 @@ export interface ApiDeviceLocationFixup {
+} & SettingsBase;
+
 export interface ApiDeviceHistorySettings {
+  location?: LatLng;
```
I assume this would only be different from the `DeviceHistory.location` field in the `synced === false` case?

hardiesoft commented on browse-next/src/views/DeviceDiagnosticsSubView.vue:
```diff
@@ -46,11 +63,137 @@ const deviceConfig = ref<LoadedResource<DeviceConfigDetail>>(null);
+  return null;
+};
+
+const handleToggleUseLowPowerMode = async (setting: { name: string }) => {
```
With all of these toggles, there needs to be some kind of "change in progress" variable that disables the button while the change is happening, otherwise users will be able to mash the buttons while the request is in flight.

hardiesoft commented on browse-next/src/views/DeviceDiagnosticsSubView.vue:
```diff
@@ -46,11 +63,137 @@ const deviceConfig = ref<LoadedResource<DeviceConfigDetail>>(null);
+  }
+  const windows = settings.value.windows;
+  customSettings.value = {
+    powerOn: windows.powerOn,
```
As noted elsewhere, I think we should probably remove `powerOn/Off`

hardiesoft commented on browse-next/src/views/DeviceDiagnosticsSubView.vue:
```diff
@@ -454,7 +767,123 @@ const deviceLocationPoints = computed<NamedPoint[]>(() => {
+          <template #cell(actions)="row">
+            <div v-if="row.item.name === 'Use Low Power Mode'">
+              <b-button
+                @click="handleToggleUseLowPowerMode(row.item)"
```
Toggling low power mode should only be available on devices that support it.  One way to work out what kind of device we're looking at currently is to check the `saltNodeGroup` variable on this page for a substring of `tc2`.

hardiesoft commented on browse-next/src/views/DeviceDiagnosticsSubView.vue:
```diff
@@ -454,7 +767,123 @@ const deviceLocationPoints = computed<NamedPoint[]>(() => {
       </div>
     </div>
     <div class="mt-4" v-if="device.type === 'thermal'">
+      <h6 class="mt-4">Device Settings:</h6>
```
This really needs a different view for mobile responsive breakpoints – however, I'm also inclined to want to move this all to the Setup tab, since it seems more appropriate there, so I can take care of that if you want.

hardiesoft commented on browse-next/src/views/DeviceDiagnosticsSubView.vue:
```diff
@@ -454,7 +767,123 @@ const deviceLocationPoints = computed<NamedPoint[]>(() => {
+                >24 Hours</b-button
+              >
+              <b-button
+                @click="enableCustomRecordingWindows"
```
In my testing, `enableCustomRecordingWindows` did nothing, I assume it's because there were not already window settings defined in my device settings, so it returned early?  That case probably wants to be handled.

hardiesoft reviewed (changes requested):
Hey Patrick, thanks for this.  Looks mostly pretty good, but I did find a few things that didn't work quite right, so let me know when you've addressed those issues and I can re-review :)

hardiesoft committed:
- Add a script to fixup device history settings corruption by ratThresh script, and fix ratThresh script so that it doesn't nuke settings.

hardiesoft committed:
- Make ratThresh script only update device settings if there are new human tagged rodents.

hardiesoft committed:
Merge branch 'feat-sync-settings' of https://github.com/TheCacophonyProject/cacophony-web into feat-sync-settings

hardiesoft committed:
- Make ratThresh script only update device settings if there are new human tagged rodents (2)

Zainrax committed:
remove loction from device history settings. Remove poweron/off for tc2

hardiesoft committed:
- Adding test coverage for device-settings

hardiesoft committed:
Merge branch 'feat-sync-settings' of https://github.com/TheCacophonyProject/cacophony-web into feat-sync-settings

# Conflicts:
#	api/api/V1/Device.ts

hardiesoft committed:
- Make sure location specific device settings get purged when creating a new station for a device, while previously set device recording window and recording settings get retained.

Zainrax committed:
minor fix on settings api, handle no settings in browse next

hardiesoft committed:
- Fix tagged recording query
- Set lastConnectionTime to null if newer recordings are uploaded via sidekick so that we see a device as being "offline" from that point.
- Fix device settings update in place issue.

hardiesoft committed:
Merge branch 'feat-sync-settings' of https://github.com/TheCacophonyProject/cacophony-web into feat-sync-settings

# Conflicts:
#	api/api/V1/Device.ts
#	api/models/DeviceHistory.ts

hardiesoft committed:
- Sync changes with upstream.

hardiesoft committed:
- Remove ratthreshold debug code

hardiesoft committed:
- Fix UI tests

hardiesoft committed:
- Moving settings sync UI to device setup tab (WIP)
- Adding flag to get latest *synced* settings
- Add tests to confirm/sync settings, and to test getting latest synced settings.

hardiesoft committed:
- Moving settings sync UI to device setup tab
- Adding flag to get latest *synced* settings
- Add latest battery info to devices listing, device diagnostic page, and link to device page from recording view.

hardiesoft committed:
- Moving settings sync UI to device setup tab
- Adding flag to get latest *synced* settings
- Add latest battery info to devices listing, device diagnostic page, and link to device page from recording view.

hardiesoft committed:
- Add spinner while saving window settings
